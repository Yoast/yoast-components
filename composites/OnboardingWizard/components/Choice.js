import React from "react";
import Input from "../../../forms/Input";
import Label from "../../../forms/Label";
import htmlDecoder from "../helpers/htmlDecoder";
import Explanation from "./Explanation";

/**
 * Represents a choice interface, like a group of radio buttons or a select button. It could render a
 * group of radio buttons (default) or a selectbox
 *
 * @param {Object} props The properties.
 * @returns {JSX} The choice component.
 * @constructor
 */
const Choice = ( props ) => {
	let choices = props.properties.choices;
	let fieldKeys = Object.keys( choices );
	let fieldName = props.name;
	let type = props.properties.type;

	if ( typeof type === "undefined" ) {
		type = "radio";
	}
	let wrapperClass = "yoast-wizard-input-" + type;

	let fieldSet = () => {
		let field;

		if ( type === "select" ) {
			field =
				<fieldset className={"yoast-wizard-input-select-" + fieldName}>
					<select defaultValue={props.value} name={fieldName}
					        className={props.optionClassName} onChange={props.onChange}>
						{fieldKeys.map( ( choiceName, index ) => {
							let choice = choices[ choiceName ];

							return (
								<option value={choiceName} key={index}>
									{htmlDecoder( choice.label )}
								</option>
							);
						} )}
					</select>
				</fieldset>
			;
		} else {
			field =
				<fieldset className={"yoast-wizard-input-radio-" + fieldName}>
					{fieldKeys.map( ( choiceName, index ) => {
						let choice = choices[ choiceName ];
						let id = `${fieldName}-${index}`;
						// If the value for the choice field equals the name for this choice, the choice is checked.
						let checked = ( props.value === choiceName );

						return (
							<div className={props.optionClassName + " " + choiceName} key={index}>
								<Input name={fieldName} type="radio" label={choice.label} onChange={props.onChange}
								       value={choiceName} optionalAttributes={ { id, checked } }
								/>
								<Label for={id}
								       optionalAttributes={ { "aria-label": choice.screenReaderText } }>{htmlDecoder( choice.label )}</Label>
							</div>
						);
					} )}
				</fieldset>
			;
		}

		return field;
	};

	return (
		<div className={wrapperClass}>
			<p className="yoast-wizard-field-description">{props.properties.label}</p>
			{fieldSet()}
			<Explanation text={props.properties.explanation}/>
		</div>
	);
};

Choice.propTypes = {
	component: React.PropTypes.string,
	type: React.PropTypes.string,
	value: React.PropTypes.string,
	properties: React.PropTypes.object,
	"default": React.PropTypes.string,
	name: React.PropTypes.string.isRequired,
	onChange: React.PropTypes.func,
	className: React.PropTypes.string,
	optionClassName: React.PropTypes.string,
};

Choice.defaultProps = {
	component: "",
	type: "radio",
	value: "",
	properties: {
		label: "",
		choices: {},
	},
	"default": "",
};

export default Choice;
