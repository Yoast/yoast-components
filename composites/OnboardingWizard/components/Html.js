import React from "react";

/**
 *  Represents a piece of HTML. This can be useful on the opening and closing screen of the wizard, to add some
 *  introduction text, a success message or CTA towards the end.
 *
 * @param {Object} props The properties to use.
 * @returns {JSX} The HTML component.
 * @constructor
 */
const HTML = ( props ) => {
	return (
		<div dangerouslySetInnerHTML={{ __html: props.properties.html }}></div>
	);
};

HTML.propTypes = {
	properties: React.PropTypes.object.isRequired,
};

HTML.defaultProps = {
	properties: {
		html: "",
	},
};

export default HTML;
