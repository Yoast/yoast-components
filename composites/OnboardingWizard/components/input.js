import React from "react";

/**
 * Represents a text input interface, like a regular input field or a text area. Initially it should render a
 * normal text input. We might add other representations later on.
 *
 * @param {Object} props The properties to use.
 * @returns {JSX} The input component.
 * @constructor
 */
const Input = ( props ) => {
	return (
		<div>
			<h2>{props.properties.label}</h2>
			<label htmlFor={props.fieldName}>{props.label}</label>
			<input onChange={props.onChange} type={props.type} id={props.fieldName} name={props.fieldName}
			       defaultValue={props.data} placeholder={props.placeholder}/>
		</div>
	);
};

Input.propTypes = {
	label: React.PropTypes.string.isRequired,
	placeholder: React.PropTypes.string,
	type: React.PropTypes.string.isRequired,
	name: React.PropTypes.string.isRequired,
	value: React.PropTypes.string,
	properties: React.PropTypes.shape(
		{
			label: React.PropTypes.string.isRequired,
		}
	),
	data: React.PropTypes.string,
	onChange: React.PropTypes.func,
	fieldName: React.PropTypes.string.isRequired,
};

Input.defaultProps = {
	label: "input field: ",
	placeholder: "enter text..",
	type: "text",
	name: "input",
	data: "",
	fieldName: "",
};

export default Input;
