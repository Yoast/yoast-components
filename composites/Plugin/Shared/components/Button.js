import React from "react";
import styled from "styled-components";
import flow from "lodash/flow";
import omit from "lodash/omit";
import PropTypes from "prop-types";

import colors from "../../../../style-guide/colors.json";
import SvgIcon from "./SvgIcon";
import { rgba } from "../../../../style-guide/helpers";
import { getRtlStyle } from "../../../../utils/helpers/styled-components";

const settings = {
	minHeight: 32,
	verticalPadding: 4,
	borderWidth: 1,
};

const ieMinHeight = settings.minHeight - ( settings.verticalPadding * 2 ) - ( settings.borderWidth * 2 );

/**
 * Returns a component with applied base button styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied base button styles.
 */
export function addBaseStyle( component ) {
	return styled( component )`
		display: inline-flex;
		align-items: center;
		justify-content: center;
		vertical-align: middle;
		border-width: ${ `${ settings.borderWidth }px` };
		border-style: solid;
		margin: 0;
		padding: ${ `${ settings.verticalPadding }px` } 10px;
		border-radius: 3px;
		cursor: pointer;
		box-sizing: border-box;
		font-size: inherit;
		font-family: inherit;
		font-weight: inherit;
		text-align: ${ getRtlStyle( "left", "right" ) };
		overflow: visible;
		min-height: ${ `${ settings.minHeight }px` };

		svg {
			// Safari 10
			align-self: center;
		}

		// Only needed for IE 10+. Don't add spaces within brackets for this to work.
		@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
			::after {
				display: inline-block;
				content: "";
				min-height: ${ `${ ieMinHeight }px` };
			}
		}
	`;
}

/**
 * Returns a component with applied focus styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied focus styles.
 */
export function addFocusStyle( component ) {
	return styled( component )`
		&::-moz-focus-inner {
			border-width: 0;
		}

		&:focus {
			outline: none;
			border-color: ${ props => props.focusBorderColor };
			color: ${ props => props.focusColor };
			background-color: ${ props => props.focusBackgroundColor };
			box-shadow: 0 0 3px ${ rgba( colors.$color_blue_dark, .8 ) };
		}
	`;
}

/**
 * Returns a component with applied hover styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied hover styles.
 */
export function addHoverStyle( component ) {
	return styled( component )`
		&:hover {
			color: ${ props => props.hoverColor };
			background-color: ${ props => props.hoverBackgroundColor };
			border-color: ${ props => props.hoverBorderColor };
		}
	`;
}

/**
 * Returns a component with applied active styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied active styles.
 */
export function addActiveStyle( component ) {
	return styled( component )`
		&:active {
			color: ${ props => props.activeColor };
			background-color: ${ props => props.activeBackgroundColor };
			border-color: ${ props => props.activeBorderColor };
			box-shadow: inset 0 2px 5px -3px ${ rgba( colors.$color_button_border_active, 0.5 ) };
		}
	`;
}

/**
 * Returns a component with applied font size style.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied font size styles.
 */
export function addFontSizeStyles( component ) {
	return styled( component )`
		font-size: 0.8rem;
	`;
}

/**
 * Returns a component with all button selector styles applied.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied styles.
 */
export const addButtonStyles = flow( [
	/*
	 * Styled-components applies the generated CSS classes in a reversed order,
	 * but we want them in the order: base - hover - focus - active.
	 */
	addActiveStyle,
	addFocusStyle,
	addHoverStyle,
	addBaseStyle,
] );

/**
 * Returns a basic styled button.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} styled button.
 */
export const BaseButton = addButtonStyles(
	styled.button`
		color: ${ props => props.textColor };
		border-color: ${ props => props.borderColor };
		background: ${ props => props.backgroundColor };
		box-shadow: 0 1px 0 ${ props => rgba( props.boxShadowColor, 1 ) };
	`
);

BaseButton.propTypes = {
	type: PropTypes.string,
	backgroundColor: PropTypes.string,
	textColor: PropTypes.string,
	borderColor: PropTypes.string,
	boxShadowColor: PropTypes.string,
	hoverColor: PropTypes.string,
	hoverBackgroundColor: PropTypes.string,
	hoverBorderColor: PropTypes.string,
	activeColor: PropTypes.string,
	activeBackgroundColor: PropTypes.string,
	activeBorderColor: PropTypes.string,
	focusColor: PropTypes.string,
	focusBackgroundColor: PropTypes.string,
	focusBorderColor: PropTypes.string,
};

BaseButton.defaultProps = {
	type: "button",
	backgroundColor: colors.$color_button,
	textColor: colors.$color_button_text,
	borderColor: colors.$color_button_border,
	boxShadowColor: colors.$color_button_border,
	hoverColor: colors.$color_button_text_hover,
	hoverBackgroundColor: colors.$color_button_hover,
	hoverBorderColor: colors.$color_button_border_hover,
	activeColor: colors.$color_button_text_hover,
	activeBackgroundColor: colors.$color_button,
	activeBorderColor: colors.$color_button_border_hover,
	focusColor: colors.$color_button_text_hover,
	focusBackgroundColor: colors.$color_white,
	focusBorderColor: colors.$color_blue,
};

/**
 * Returns a styled Button with set font size.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} styled button.
 */
export const Button = addFontSizeStyles( BaseButton );

/**
 * Applies styles to SvgIcon for IconButton with text.
 *
 * @param {ReactElement} icon The original SvgIcon.
 *
 * @returns {ReactElement} SvgIcon with text styles applied.
 */
function addIconTextStyle( icon ) {
	return styled( icon )`
		margin: ${ getRtlStyle( "0 8px 0 0", "0 0 0 8px" ) };
		flex-shrink: 0;
	`;
}

/**
 * Returns an icon button that can optionally contain text.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} styled icon button.
 */
export const IconButton = ( props ) => {
	const { children: text, icon, iconColor } = props;

	let IconComponent = SvgIcon;
	if ( text ) {
		IconComponent = addIconTextStyle( IconComponent );
	}

	const newProps = omit( props, "icon" );

	return (
		<Button { ...newProps }>
			<IconComponent icon={ icon } color={ iconColor } />
			{ text }
		</Button>
	);
};

IconButton.propTypes = {
	icon: PropTypes.string.isRequired,
	iconColor: PropTypes.string,
	children: PropTypes.oneOfType( [
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
		PropTypes.string,
	] ),
};

IconButton.defaultProps = {
	iconColor: "#000",
};

/**
 * Returns an icons button that can optionally contain a prefix and / or a suffix icon.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} styled icon button.
 */
export const IconsButton = ( props ) => {
	const {
		children,
		className,
		prefixIcon,
		suffixIcon,
		...buttonProps
	} = props;

	return (
		<Button className={ className } { ...buttonProps }>
			{ prefixIcon &&
				<SvgIcon
					icon={ prefixIcon.icon }
					color={ prefixIcon.color }
					size={ prefixIcon.size }
				/>
			}
			{ children }
			{ suffixIcon &&
				<SvgIcon
					icon={ suffixIcon.icon }
					color={ suffixIcon.color }
					size={ suffixIcon.size }
				/>
			}
		</Button>
	);
};

IconsButton.propTypes = {
	className: PropTypes.string,
	prefixIcon: PropTypes.shape( {
		icon: PropTypes.string,
		color: PropTypes.string,
		size: PropTypes.string,
	} ),
	suffixIcon: PropTypes.shape( {
		icon: PropTypes.string,
		color: PropTypes.string,
		size: PropTypes.string,
	} ),
	children: PropTypes.oneOfType( [
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
		PropTypes.string,
	] ),
};

IconsButton.defaultProps = {
	prefixIcon: {
		icon: "",
		color: colors.$black,
		size: "16px",
	},
	suffixIcon: {
		icon: "",
		color: colors.$black,
		size: "16px",
	},
};
