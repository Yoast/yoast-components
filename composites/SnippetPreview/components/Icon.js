import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import * as Icons from "../../../style-guide/svg";

/**
 * Returns an icon.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Icon.
 */
export const Icon = ( props ) => {
	const IconComponent = styled( Icons[ props.icon ] )`
		float: left;
		width: ${ props => props.size }px;
		height: ${ props => props.size }px;
		fill: ${ props => props.color };
	`;
	return <IconComponent { ...props } />;
};

Icon.propTypes = {
	icon: PropTypes.string,
	color: PropTypes.string,
	size: PropTypes.number,
};

Icon.defaultProps = {
	size: 16,
	color: "#555",
	icon: "edit",
};
