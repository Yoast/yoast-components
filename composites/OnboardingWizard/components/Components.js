import Choice from "../../Choice/Choice";
import Input from "../../../forms/Input";
import HTML from "./Html";
import MailchimpSignup from "./custom_components/MailchimpSignup";
import PublishingEntity from "./custom_components/PublishingEntity";
import PostTypeVisibility from "./custom_components/PostTypeVisibility";
import ConnectGoogleSearchConsole from "./custom_components/ConnectGoogleSearchConsole";

let Components = {
	Choice,
	Input,
	HTML,
	MailchimpSignup,
	PostTypeVisibility,
	ConnectGoogleSearchConsole,
	PublishingEntity,
};

export default Components;
