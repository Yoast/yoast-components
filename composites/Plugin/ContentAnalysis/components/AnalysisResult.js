import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import colors from "../../../../style-guide/colors.json";

import { circle } from "../../../../style-guide/svg/index";
import { eye } from "../../../../style-guide/svg/index";
import { Icon } from "../../Shared/components/Icon.js";
import IconButtonToggle from "../../Shared/components/IconButtonToggle.js";

/**
 * Returns a AnalysisResultBase component.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} The rendered AnalysisResultBase component.
 */
const AnalysisResultBase = styled.li`
	// This is the height of the IconButtonToggle.
	min-height: 24px;
	padding: 4px 4px;
	display: inline-flex;
	align-items: center;
`;

const AnalysisResultText = styled.p`
	margin: 0 8px;
`;

export const AnalysisResult = ( props ) => {
	return (
		<AnalysisResultBase { ...props }>
			<Icon
				icon={ circle }
				color={ props.bulletColor }
				size="13px"
				aria-label={ "SEO score " + props.score }
				aria-hidden={ false }
				focusable={ true }
			/>
			<AnalysisResultText>{ props.resultText }</AnalysisResultText>
			{
				props.eyeButtonIsActive
					? <IconButtonToggle
						onClick={ props.onClick }
						id={ props.id }
						icon={ eye }
						pressed={ props.pressed }
						ariaLabel={ props.ariaLabel }
					/>
					: null
			}
		</AnalysisResultBase>
	);
};

AnalysisResult.propTypes = {
	resultText: PropTypes.string.isRequired,
	bulletColor: PropTypes.string.isRequired,
	eyeButtonIsActive: PropTypes.bool.isRequired,
	id: PropTypes.string.isRequired,
	pressed: PropTypes.bool.isRequired,
	ariaLabel: PropTypes.string.isRequired,
	onClick: PropTypes.func.isRequired,
	score: PropTypes.string.isRequired,
};

export default AnalysisResult;
