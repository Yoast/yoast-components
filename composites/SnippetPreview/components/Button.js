import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import colors from "../../../style-guide/colors.json";
import { Icon } from "./Icon";
import { rgba } from "../../../style-guide/helpers";

/**
 * Returns a component with applied focus styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied focus styles.
 */
export function addFocusStyle( component ) {
	return styled( component )`
		&::-moz-focus-inner {
			border-width: 0;
		}

		&:focus {
			outline: none;
			border-color: ${ colors.$color_blue };
			background-color: ${ colors.$color_white };
			box-shadow: 0 0 3px ${ rgba( `${ colors.$color_blue_dark }`, .8 ) };
		}
	`;
}

/**
 * Returns a basic styled button.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const BaseButton = addFocusStyle(
	styled.button`
		float: left;
		display: inline-block;
		padding: 8px 10px;
		border: 1px solid ${ colors.$color_button_border };
		border-radius: 4px;
		background: ${ colors.$color_button };
		color: ${ colors.$color_button_text };
		cursor: pointer;
		box-sizing: border-box;
		font-size: inherit;
		font-family: inherit;
		font-weight: inherit;
		min-height: 33px;
	`
);

BaseButton.defaultProps = {
	type: "button",
};

/**
 * Returns a button styled for the SnippetPreview.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const SnippetPreviewButton = styled( BaseButton )`
	line-height: 15px;
	font-size: 0.8rem;
	
	> span {
		margin-left: 8px;
	}
`;

/**
 * Returns an icon button that can optionally contain text.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled icon button.
 */
export const IconButton = ( props ) => {
	const { children, icon, iconColor } = props;
	return (
		<SnippetPreviewButton { ...props } >
			<Icon icon={ icon } color={ iconColor } />
			{ children ? <span>{ children }</span> : null }
		</SnippetPreviewButton>
	);
};

IconButton.propTypes = {
	icon: PropTypes.string,
	iconColor: PropTypes.string,
	children: PropTypes.string,
};

IconButton.defaultProps = {
	icon: "edit",
};

