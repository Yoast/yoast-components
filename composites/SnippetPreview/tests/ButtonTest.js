import React from "react";
import renderer from "react-test-renderer";

import { BaseButton, SnippetPreviewButton } from "../components/Button";

test( "the Button matches the snapshot", () => {
	const component = renderer.create(
		<BaseButton>ButtonValue</BaseButton>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "Button executes callback", () => {
	const component = renderer.create(
		<BaseButton onClick={
			() => {
				return "clicked";
			}
		}>ButtonValue</BaseButton>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();

	tree.props.onClick();

	tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "the SnippetPreviewButton matches the snapshot", () => {
	const component = renderer.create(
		<SnippetPreviewButton>ButtonValue</SnippetPreviewButton>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "SnippetPreviewButton executes callback", () => {
	const component = renderer.create(
		<SnippetPreviewButton onClick={
			() => {
				return "clicked";
			}
		}>ButtonValue</SnippetPreviewButton>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();

	tree.props.onClick();

	tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
