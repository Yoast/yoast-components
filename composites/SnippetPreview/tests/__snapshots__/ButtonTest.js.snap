// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button executes callback 1`] = `
<button
  className="sc-bwzfXH kGsEoO sc-bdVaJa cNDoHL"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</button>
`;

exports[`Button executes callback 2`] = `
<button
  className="sc-bwzfXH kGsEoO sc-bdVaJa cNDoHL"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</button>
`;

exports[`SnippetPreviewButton executes callback 1`] = `
<button
  className="sc-htpNat bfvzQm sc-bwzfXH kGsEoO sc-bdVaJa cNDoHL"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</button>
`;

exports[`SnippetPreviewButton executes callback 2`] = `
<button
  className="sc-htpNat bfvzQm sc-bwzfXH kGsEoO sc-bdVaJa cNDoHL"
  onClick={[Function]}
  type="button"
>
  ButtonValue
</button>
`;

exports[`the Button matches the snapshot 1`] = `
<button
  className="sc-bwzfXH kGsEoO sc-bdVaJa cNDoHL"
  type="button"
>
  ButtonValue
</button>
`;

exports[`the SnippetPreviewButton matches the snapshot 1`] = `
<button
  className="sc-htpNat bfvzQm sc-bwzfXH kGsEoO sc-bdVaJa cNDoHL"
  type="button"
>
  ButtonValue
</button>
`;
